### PR TITLE
chore(node): enforce encoding parameters

### PR DESCRIPTION
**Pull request by SimonSchick**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://nodejs.org/api/buffer.html
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

This will likely have breaking changes but will vastly improve type safety, we may not merge this until node v12 lands.

tl;dr: Replace occurences of `string` with `BufferEncoding` where context is encoding related, exception is TextEncoder/Decoder.

### CODE REVIEW

**Comment by SimonSchick** [comment]
@Flarna feel free to peek, I probably missed something :)

**Comment by SimonSchick** [comment]
One notable issue, NodeJS is rather lean on encoding, it apparently checks encoding case insensitive and also dashes in some places eg:
```
> Buffer.from('53', 'hex').toString('UtF-8')
'S'
```

I went ahead and normalized all occurences to `utf8`, but I noticed that `utf-8` was used in a couple of places, should we add to retain compability?

**Comment by Flarna** [comment]
node supports a lot encoding variants, see https://github.com/nodejs/node/blob/master/lib/internal/util.js#L107

**Comment by SimonSchick** [comment]
@Flarna yea I just saw that, I don't think it's reasonable to add all of those, I think we should stick with: https://nodejs.org/docs/latest-v11.x/api/all.html#buffer_buffers_and_character_encodings
This will help streamlining the usage of encoding on a global scale.

**Comment by Flarna** [comment]
I agree but I fear if this PR gets merged a lot projects will break as they use `string`. Syncing with Node 12 will not help as a lot other modules are also effected.

**Comment by SimonSchick** [comment]
@rbuckton any thoughts on this?
Also can you teach the TS bot not to be as aggressive on node PRs?
They tend to be a little bit long running in general.

**Comment by SimonSchick** [comment]
@Flarna @amcasey could I get some feedback if this kind of change is desired at all before I re-open this PR?
Otherwise I will reopen it as a part of v12.

**Comment by Flarna** [comment]
I agree that it improves type safety but based on how much other definitions you had to touch it's quite sure that this will break quite a lot projects. Not sure if it is worth to do this as I'm quite sure there will be a lot complains because of breaking changes in a patch release - as there is no semver versioning for type definitions.

**Comment by amcasey** [comment]
+1 This is a great way to improve intellisense and eliminate errors, but it's probably too disruptive to justify those benefits.
